p2p/simulations/events_test.go

```diff
@@ -0,0 +1,166 @@
+package simulations
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p/discover"
+	"github.com/ethereum/go-ethereum/p2p/simulations/adapters"
+)
+
+func init() {
+	log.Root().SetHandler(log.LvlFilterHandler(log.LvlInfo, log.StreamHandler(os.Stderr, log.TerminalFormat(true))))
+}
+
+// This test confirms (and demonstrates) using simulation events
+// to determine whether starting and connecting nodes
+//
+func TestNodeUpAndConn(t *testing.T) {
+
+	adapter := adapters.NewSimAdapter(adapters.Services{
+		"test": newTestService,
+	})
+	network := NewNetwork(adapter, &NetworkConfig{
+		DefaultService: "test",
+	})
+	nodeCount := 3
+	ids := make([]discover.NodeID, nodeCount)
+
+	for i := 0; i < nodeCount; i++ {
+		conf := adapters.RandomNodeConfig()
+		node, err := network.NewNodeWithConfig(conf)
+		ids[i] = node.ID()
+		if err != nil {
+			t.Fatalf("error creating node: %s", err)
+		}
+	}
+
+	quitC := make(chan struct{})
+	trigger := make(chan discover.NodeID)
+	events := make(chan *Event)
+	sub := network.Events().Subscribe(events)
+	defer sub.Unsubscribe()
+
+	action := func(ctx context.Context) error {
+		go func() {
+			for {
+				select {
+				case ev := <-events:
+					if ev == nil {
+						panic("got nil event")
+					} else if ev.Type == EventTypeNode {
+						if ev.Node.Up {
+							log.Info("got node up event", "event", ev, "node", ev.Node.Config.ID)
+							trigger <- ev.Node.Config.ID
+						}
+					}
+
+				case <-quitC:
+					log.Warn("got quit action 1")
+					return
+				}
+
+			}
+		}()
+		go func() {
```

Comment:
Why is this in a go-routine? It might as well be plainly in the `action` func main goroutine.